File: malloc_copy/malloc_copy.js
/**
 * @author <code@tythos.net>
 */

require(['../lib/logger', '../lib/wasi.esm'], function(logger, wasi) {
  (async () => {
    const memory = new WebAssembly.Memory({initial: 2});
    const response = await fetch('./malloc_copy.wasm');
    const bytes = await response.arrayBuffer();
    let wmi = new wasi.WASI();
    const {instance} = await WebAssembly.instantiate(
        bytes, {'env': {memory}, 'wasi_snapshot_preview1': wmi.wasiImport});
    const text = 'Hello from JavaScript!';
    const view1 = new Uint8Array(memory.buffer);
    const pInput = instance.exports.malloc(1024);
    const view2 = new Uint8Array(memory.buffer);
    encode(view2, pInput, text);
    const pOutput = instance.exports.malloc_copy(pInput, text.length);
    logger.log(`copy: ${decode(view2, pOutput)}`);
  })();

  const encode = (memory, base, string) => {
    for (let i = 0; i < string.length; i++) {
      memory[base + i] = string.charCodeAt(i);
    }
    memory[base + string.length] = 0;
  };

  const decode = (memory, base) => {
    let cursor = base;
    let result = '';
    while (memory[cursor] !== 0) {
      result += String.fromCharCode(memory[cursor++]);
    }
    return result;
  };
});

File: malloc_copy/malloc_copy.c
/**
 * @file malloc_copy.c
 * @author Brian Kirkpatrick (code@tythos.net)
 * @brief See README.md
 * @version 0.1
 * @date 2021-11-06
 * 
 * @copyright Copyright (c) 2021
 * 
 */

#include <string.h>
#include <stdlib.h>

char* malloc_copy(char* input) {
    char* result = malloc(1024);
    strncpy(result, input, strlen(input));
    return result;
}
